---
title: "Analysis of Dominique's  RNAseq Data "
output: 
  html_notebook:
    number_sections: yes
    toc: yes
    toc_float: yes
    theme: united
    highlight: tango
    code_folding: hide
author: "Frank"
date: "`r date()`"
---
# Libraries 

```{r}
require(dplyr)
require(limma)
require(ggplot2)
```

# Read the data
```{r}
raw.count <- read.delim(file="./counts-gene.txt",header = T,skip = 1,
                        stringsAsFactors = F)
anno.count <- raw.count[,1:6]; raw.count <- raw.count[,-c(1:6)]
colnames(raw.count) <- sub(".filtered.bam","",colnames(raw.count)) 
sample.info <- read.csv(file="./sample_info.csv",stringsAsFactors = F,
                        header = T)
anno.count$Geneid <- substr(anno.count$Geneid,1,18)
```

# DE analysis (Limma + Voom method)

```{r}
all.equal(colnames(raw.count),paste0("n",sample.info$Number)) # TRUE 
sample.info$Number <- colnames(raw.count)
sample.info$treat<-with(sample.info,paste(Genotype,X_CD3.CD28..hrs.,X_4.1BB..hrs.,sep = "_"))
sample.info$treat <- factor(sample.info$treat)
rownames(sample.info)<- sample.info$Number
sample.info$rep <- paste0('r',rep(c(1,2,3),each=6))
sample.info$key <- paste(sample.info$treat,sample.info$rep,sep= '_')
colnames(raw.count) <- sample.info[colnames(raw.count),'key']
# check the mapped lib size 
barplot(apply(raw.count,2,sum)/10^6)


# rpkm calulation 
require(edgeR)
keep <- rowSums(raw.count)>0
raw.count.2 <- raw.count[keep,]
anno.count.2 <- anno.count[keep,]
all.rpkm <- rpkm(raw.count.2,gene.length = anno.count.2$Length)

lib.size <- colSums(raw.count.2)
raw.count.2[1,'NES_24_24_r2']/lib.size['NES_24_24_r2']*10^6/anno.count.2$Length[1]*1000 == all.rpkm[1,'NES_24_24_r2']

anno.count.2$Symbol = getGsymble(anno.count.2$Geneid)
write.csv(file='rpkm_expressed_gene.csv',
          cbind(anno.count.2[,c(1,6,7)],
                     all.rpkm)
          ,quote = F,row.names = F)
```
## WT 
```{r}

sample.info.sub <- subset(sample.info,Genotype=="WT")

require(edgeR)
y <- DGEList(counts = raw.count,genes = anno.count)
isExp <- rowSums(y$counts)>50
ysub <- y[isExp,,keep.lib.size=F]
dim(y)
y <- calcNormFactors(y)
plotMDS(y,labels=sample.info$treat,col=ifelse(sample.info$Genotype=="WT",1,2),
        gene.selection = "common",top=100,prior.count = 5)

design <- model.matrix(~treat+0 ,sample.info)
v <- voom(y, design, plot = TRUE)
```

```{r}
cor <- duplicateCorrelation(v, design)
fit <- lmFit(v, design)
cont.matrix <- makeContrasts(treatWT_12_12-treatWT_12_0,
                             treatWT_24_24-treatWT_12_0,
                             
                             levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2, 0.01)
topTable(fit2, coef=1, adjust="BH",number = nrow(fit2))
results <- decideTests(fit2,adjust.method = "BH",p.value = 0.01,lfc = 1)
vennDiagram(results)
```


# DE analysis (EdgeR method )
```{r}

```

# Session Info 
```{r}
 
```


